projecttacoma/fqm-testify
# Measure Calculation on All Patients

# Summary
Adds functionality for the user to run measure calculation on all test cases and see the population results in a tabular format.

## New Behavior
This PR corresponds to the following user story: “As a user, I want to be able to run measure calculation on my test case for a given measure and see a human-readable table that shows the population results for every patient I used during the calculation”

Adds a button (“Calculate Population Results”) that allows the user to run measure calculation of all the created patients for the uploaded FHIR Measure. The population results for each patient appear in a population results table, with a row for each patient, labeled with the patient’s name and date of birth. This table appears as a popover so that the user can still view the current patients and their resources.

## Code Changes
* New component - `PopulationCalculation` - with functions for gathering all the patient names/DOBs, calculating measure reports for each patient using fqm-execution, throwing errors from fqm-execution in a Mantine notification, etc.
    * “Calculate Population Results” button appears when there is at least one patient, and a measure bundle has been uploaded
    * Population Results popover appears when the calculation ends and is successful
* Updated `PatientCreation` component to be scrollable on overflow (allows for more room for displaying population results table and calculation button)
* Bring in `ecqm-visualizers` as a dependency

# Testing Guidance
Run unit tests with `npm run test`. Then, try the following in the app:
* Start the app and check that the “Calculate Population Results” button is not present
* Add at least one patient and upload a measure bundle, and then check that the “Calculate Population Results” button appears at the bottom of the page, below the resource creation panel
* Try creating patients that will fall into the numerator/denominator/IPP populations (I found it easiest to test this by uploading the EXM130 measure bundle, then uploading the EXM130 numerator/denominator patient bundles and making sure the measurement period is for 1/1/2019-12/31/2019)
* After adding some patients, click the “Calculate Population Results” button and check that the table renders correctly in the popover
* Check that the popover can be closed when clicking on the “Close” button

### Other Notes
The way in which we display the population results table can be changed. We want to be able to still view the current patients while having the table on the screen, so we don’t want to route to a new page, and displaying the table at the bottom of the screen may make it hard for the user to reference the patients while looking at the results table. I’m not super happy with the popover, so I am open to suggestions on how best to display the table, especially when there is a large number of patients that are performing calculation on.
